qiskit/transpiler/passes/utils/minimum_point.py

```diff
@@ -0,0 +1,93 @@
+
+    Fields used by this pass in the property set are (all relative to the ``prefix``
+    argument):
+
+     * ``{prefix}_minimum_point_count`` - Used to track the number of executions since
+        the current minimum was found
+     * ``{prefix}_backtrack_history`` - Stores the current minimum value and :class:`~.DAGCircuit`
+     * ``{prefix}_minimum_point`` - This value gets set to ``True`` when either a fixed point
+        is reached over the ``backtrack_depth`` executions, or ``backtrack_depth`` was exceeded
+        and an earlier minimum is restored.
+    """
+
+    def __init__(self, property_set_list, prefix, backtrack_depth=5):
+        """Initialize an instance of this pass
+
+        Args:
+            property_set_list (list): A list of property set keys that will
+                be used to evaluate the local minimum. The values of these
+                property set keys will be used as a tuple for comparison
+            prefix (str): The prefix to use for the property set key that is used
+                for tracking previous evaluations
+            backtrack_depth (int): The maximum number of entries to store. If
+                this number is reached and the next iteration doesn't have
+                a decrease in the number of values the minimum of the previous
+                n will be set as the output dag and ``minimum_point`` will be set to
+                ``True`` in the property set
+        """
+        super().__init__()
+        self.property_set_list = property_set_list
+        self.backtrack_name = f"{prefix}_minimum_point_count"
+        self.backtrack_min_name = f"{prefix}_backtrack_history"
+        self.minimum_reached = f"{prefix}_minimum_point"
+        self.backtrack_depth = backtrack_depth
+        self.minimum_dag = None
+
+    def run(self, dag):
+        """Run the DAGFixedPoint pass on `dag`."""
+        score = tuple(self.property_set[x] for x in self.property_set_list)
+
+        if self.property_set[self.backtrack_name] is None:
+            self.property_set[self.backtrack_name] = 1
+        elif self.property_set[self.backtrack_min_name] is None:
+            self.property_set[self.backtrack_min_name] = (score, deepcopy(dag))
+        elif score > self.property_set[self.backtrack_min_name][0]:
+            self.property_set[self.backtrack_name] += 1
+            if self.property_set[self.backtrack_name] == self.backtrack_depth:
+                self.property_set[self.minimum_reached] = True
+                return self.property_set[self.backtrack_min_name][1]
+        elif score < self.property_set[self.backtrack_min_name][0]:
+            self.property_set[self.backtrack_name] = 1
+            self.property_set[self.backtrack_min_name] = (score, deepcopy(dag))
+        elif score == self.property_set[self.backtrack_min_name][0]:
+            self.property_set[self.minimum_reached] = True
+            return self.property_set[self.backtrack_min_name][1]
```

Comment:
Maybe it would be a shade cleaner to factor out all the different properties into something like a
```python
@dataclasses.dataclass
class _MinimumPointState:
    dag: DAGCircuit
    score: Tuple[float, ...]
    since: int
```
and store only that during the run?  I'm having trouble parsing all the different `self.property_set[self.backtrack_<x>]` stuff into the algorithm in my head.  The dataclass doesn't really solve that, it just removes a little bit of the boiler plate and gives things shorter names.

Then the loop of the pass would look something more like
```python
state = self.property_set[self.backtrack_state]
if state is None:
    state = self.property_set[self.backtrack_state] = _MinimumPointState(
        since=0,
        score=(math.inf,) * len(self.property_set_list),
        dag=None,
    )
if score < state.score:
    state.since = 1
    state.score = score
    state.dag = deepcopy(dag)
    return dag
if score > state.score:
    state.since += 1
if score == state.score or state.since == self.backtrack_depth:
    self.property_set[self.minimum_reached] = True
    return state.dag
return dag
```

I'm not certain I've got all the logic exactly correct there (because I find the current iteration a bit hard to parse), but certainly I think it's a shade easier to read.